### PR TITLE
fix: single and updated source of Multi protocol definitions

### PR DESCRIPTION
**Pull request by mha1**
Fixes https://github.com/EdgeTX/edgetx/issues/3427, fixes https://github.com/EdgeTX/edgetx/issues/3480

Summary of changes:
- Put all Multi protocol definitions into a consistent single source. Actually two .h files, one for the protocol list and names, the other for subtypes and capabilities, see comments in .h files as to why
- updated available protocols reflecting 1.3.3.20 (15+ protocols) 
- fixed Companion always showing f/s and channel map options in model setup disregarding the protocol capability

### CODE REVIEW

**Comment by mha1** [comment]
seems the MSYS build system has a problem: https://discord.com/channels/839849772864503828/1091503585159823432/1092607325480681593

**Comment by elecpower** [comment]
Ubuntu 20.04 Companion runtime error:
ASSERT: "rd.maxSubtype + 1 == (unsigned int) rd.protocols.length()" in file /src/edgetx/companion/src/firmwares/multiprotocols.h, line 68

**Comment by pfeerick** [comment]
Looks like https://github.com/EdgeTX/edgetx/pull/3437 aims to fix the build issue.

File: radio/src/MultiProtoDefs.h
Comment (pfeerick) [inline radio/src/MultiProtoDefs.h:24]: typos: 
Common list of supported Multi protocols (currently **reflecting Multi** firmware 1.3.320)

File: radio/src/MultiProtoDefs.h
Comment (pfeerick) [inline radio/src/MultiProtoDefs.h:19]: Please use the header in  `tools/copyright-header.txt`

File: radio/src/MultiSubtypeDefs.h
Comment (pfeerick) [inline radio/src/MultiSubtypeDefs.h:19]: Please use the header in  `tools/copyright-header.txt`

**Review by pfeerick — CHANGES_REQUESTED** [review]
This is looking great so far... one stop shop has been sorely needed to help make Companion a first class citizen for MPM as well :) Just some cosmetic changes so far for me from simply looking at the code.

File: radio/src/MultiProtoDefs.h
Comment (mha1) [reply radio/src/MultiProtoDefs.h]: done

File: radio/src/MultiProtoDefs.h
Comment (mha1) [reply radio/src/MultiProtoDefs.h]: done

File: radio/src/MultiSubtypeDefs.h
Comment (mha1) [reply radio/src/MultiSubtypeDefs.h]: done

**Comment by mha1** [comment]
@elecpower found and hopefully fixed the assert problem. I wonder why Windows didn't show this.

File: radio/src/MultiSubtypeDefs.h
Comment (elecpower) [inline radio/src/MultiSubtypeDefs.h:238]: Should be // MODULE_SUBTYPE_MULTI_CONFIG non...

Comment (mha1) [reply radio/src/MultiSubtypeDefs.h]: sorry, copy/paste error - done

**Comment by elecpower** [comment]
> @elecpower found and hopefully fixed the assert problem. I wonder why Windows didn't show this.

No error with latest commit. Did you compile with DEBUG? Anyway moving onto testing.

**Comment by mha1** [comment]
> No error with latest commit. Did you compile with DEBUG? Anyway moving onto testing.

Well, that's it, I didn't.

**Comment by mha1** [comment]
@elecpower please repeat your tests

**Comment by ParkerEde** [comment]
> > The protocols are offset when saved and read back. For example, starting in Companion with TX16S profile:
> > ```
> > 1. set multi protocol to FrSkyX and sub-type to Cloned 8ch
> > 
> > 2. save models and settings and close file
> > 
> > 3. open saved file
> > 
> > 4. protocol FrSkyD and sub-type blank
> > 
> > 5. change to step 1
> > 
> > 6. simulate radio
> > 
> > 7. Mode MULTI ESky RF protocol blank
> > ```
> 
> Thanks, think I know where to look. Be back in 1h (hopefully)

I've also checked. Looks good to me.

**Comment by mha1** [comment]
> Looks like #3437 aims to fix the build issue - i.e. if you rebase this PR now that it's merged the error will go way.

Yes, did a rebase and it finished Windows CPN.

**Comment by elecpower** [comment]
Companion write and read back - pass
Simulator read - fail as protocol offset

**Comment by mha1** [comment]
any specifics on the simulator read test?

**Comment by elecpower** [comment]
> any specifics on the simulator read test?

TX16S simulator
Cpn: FrSkyD   Cloned
Simu: HiSky   HK310

Cpn: FrSkyX   LBT 8ch
Simu: ESky  [blank]

Cpn: DSM  Auto
Simu: DSM  Auto

Cpn: M-Link   Default
Simu: M-Link   [no label of field value]

Cpn: Futaba   SFHSS
Simu: J6 Pro   [no label of field value]

Cpn: Kyosho2   KT-17
Simu: [blank]   [no label of field value]

I only tested these.

**Comment by ParkerEde** [comment]
I tested also with TX16S profil.
I can't reproduce this what elecpower reported.
I build the companion and libedgetx-tx16s-simulator.dll.
For me it works as it should

**Comment by ParkerEde** [comment]
If I create a new file in Companion and a new model in it, I cannot reproduce the problem.
If I use an existing file from my Horus X10 Express and convert it to the TX16S profile, I can reproduce the problem. But only with the ext. MPM. With the int. MPM it runs correctly.
So if I now set the ext. module to "OFF" for the existing file with the error, save the file, then close Companion, restart it, open the file again and then make the setting for the ext. MPM again, it works correctly. @elecpower Can you test this?

**Comment by elecpower** [comment]
Ok so I deleted everything and rebuilt Companion, TX16S, X9D+ and Boxer sims.
TX16s retested all above. Boxer and x9d+ tested a couple above. Sims now match Companion.
So humble apologies for the second tests. I can only assume I overwrote the sims I built for this branch with some other work.
Still to test the covert scenario.

**Comment by elecpower** [comment]
In 2.8.1 created X10 Express model with internal DSM Auto and external  FrSkyX Cloned 8ch multi modules.

Opened in X10 Express profile. Noted DSM Auto but FrSkyX Cloned not Cloned 8ch.
Changed to TX16S profile and model converted.
Both modules converted including bug.
Closed file without saving.

Reopened file with TX16S profile and received convert message and chose convert.
Both modules converted as above including bug.

**Comment by mha1** [comment]
> 16S profile and received convert message and chose convert.
> Both modules converted as above including bug.

Hi Neil,
Thanks for testing. I have a hard time reproducing the X10 Express test case. First of all, how did you create the model? radio or Companion? And how did you manage to select Multi as internal protocol on the X10 express?

What I did:
- installed latest merge build (https://github.com/EdgeTX/edgetx/suites/12046948648/artifacts/633713572)
- start Companion 
- select x10Express radio profile
- file, new, +model -> document1.etx, wizard shows up, canceled wizard
- double click on MODEL01 in document1.etx window -> Companion Setup page comes up
- internal module ISRM, external module Multi FrySkyX Cloned 8ch
- close Setup
- file, save document1.etx
- close Companion

- restart Companion (still X10 express profile)
- file, open document1.etx
- double click on MODEL01
- setup opens with internal module ISRM, external module Multi FrySkyX Cloned 8ch
- simulate model, internal module shows ISRM, external module shows Multi FrySkyX Cloned 8ch
- close Setup
- close Companion

- restart Companion
- select TX16s profile
- file, open document1.etx
- double click on MODEL01
- setup opens with warning internal module set to OFF, external module Multi FrySkyX Cloned 8ch
- simulate model, internal module shows Off, external module shows Multi FrySkyX Cloned 8ch
- close Setup
- close Companion

So all successful. I'd appreciate posting a dummy (mha1) proof sequence on how to reproduce what you saw.

Thanks,
Michael

**Comment by mha1** [comment]
Btw: Dummy me created a major mess with #3355 which showed it's ugly face in 2.8.2 (#3434) and is also present in 2.9. My fudge up causes under very specific conditions (which makes it look random) an index shift to Multi protocols (firmware and simulator). The reason is writing to the PXX2 module data struct union without checking if PXX2 (ISRM) is actually an active module of the loaded model. This semi-randomly overwrites the Multi union of the same data struct causing an index shift in in the .yml file of the active model. I drafted a trial firmware version for the #3434 problem finders (waiting for feedback) and included this tentative fix also temporarily in this PR to stop it from potentially compromising testing.

Please wait for the merge checks to complete and use the assets generated.

Sorry for the inconvenience.

**Comment by elecpower** [comment]
I couldn't repeat in the light of a new day, so I've retested 26 protocol / sub-type combinations in Companion main branch and compared to this branch. Most were the main branch FrSky protocol which has been remapped in this branch to newer FrSky protocols / sub-types. All mapped okay. I think I confused myself on the FrSky D16 Clone to FrSkyX Cloned. 
Rule #1 do not test complicated stuff like this when tired. 

Also I configured the hardware for the X10 Express to have an internal Multi module so I could test 2 protocols per model.

Great work on this PR, well overdue. Once merged it will give me a base to clean up the dog's breakfast of a ui which is part of my feature branch I parked in the bottom draw.

**Comment by mha1** [comment]
Updated Multiprotocol database to current MPM project status (ref. MPM Multiprotos.ino)
- changed protocol name "FX816" to "FX"
- added Flysky AFHDS2A options for PPM/PWM SBUS 16ch "PWM,SB16" and  "PPM,SB16
- added V761 option "TOPRC"
- added FX options "816" and "620"
- added FRSKYR9 option "-- 8ch"

**Comment by mha1** [comment]
MPM protocol scan excludes some protocols which are not meant to be selected. For Companion and Simulator (and for radios if there is no MPM connected) this was handled in different places with outdated information. Harmonized and centralized the handling of non-selectabale protocols into the single source for all companion, Simulator and all GUI types avoiding "invalid protocols" between Companion and radio.

**Comment by mha1** [comment]
More cleanup:
- changed some MPM protocol names to the ones the MPM delivers with MPM protocol scan to harmonize names between CPN/Simulator/radio
- added XN2927DUMP and MOULDKG to the list of excluded MPM protocols just like the MPM does with MPM protocol scan
- added new option value limits and moved definitions to single source
- adapted CPN and radio to use common definitions of option values for all GUI types 
- moved distributed MPM (as far as possible) to one coherent place under a single #if defined (MULTIMODULE)
- fixed option value not reset at change of protocol
- some editorial stuff

Note: There is still some work to be done on CPN GUI option values. DSM2 servo rates are hard coded, BAYANG telemetry options are only shown by index and not by text as in radio code, AFHDS2A servo update rate is shown as index too which is to no use for users. 

I'll try to fix all this in the next session, but request to get rid of BAYANG telemetry options translations. It is much easier to define a common string list by using untranslated.h

**Comment by pfeerick** [comment]
Give them the chop - MPM options really shouldn't be translated - I can complain afterwards about it once the PR is ready to go ;)

**Review by pfeerick — APPROVED** [review]

**Comment by mha1** [comment]
chop,chop - gone

More cleanup:
- CPN: Bayang telemetry option with matching list instead of just indexes
- CPN: AFHDS2A servo frequency option with matching list instead of just indexes

I think this PR is now ready for a (re-)review now.

**Comment by elecpower** [comment]
So in testing I've found a bug (pre-existing your dabbling lol) in the Ubuntu self built X9D+ libsim of this and latest main branches, and 2.8.1 official Win 64 build where libsim crashes when changing External RF MULTI sub-type. However there appears to be no repeatable pattern. 

The bug show itself more often in self builds though it is compiled with DEBUG so obviously generated code not the same. It could also be uninitialised memory as the libsim fails faster on retrying after a crash in Ubuntu terminal at least.

Steps to reproduce:
Companion
1. X9D+ profile
2. New file
3. New Model
4. External module MULTI FlSky (does not really matter the starting entry)
5. Simulate

Radio simulator
6. Click MENU
7. Click PAGE
8. Using mouse scroll to OR click +/- until reach External RF Type
9. Click ENT
10. After a varying number of mouse scroll up and down OR click +/- and libsim crashes. There is a Segmentation Fault message in a terminal session.

The speed of change does not appear to affect the speed of crash. In some cases it has not crashed though not crashing is rare.

I haven't been able to crash a self build of Boxer libsim yet.

Would you mind seeing if you can reproduce?

**Comment by elecpower** [comment]
BTW the other changes are looking good

**Comment by mha1** [comment]
@elecpower I couldn't get X9D+ Win64 2.8.1 release and the latest Win64 check build of this PR to crash using your procedure. Just to be sure: we are going through "MODE" (module types), aren't we?

**Comment by elecpower** [comment]
> @elecpower I couldn't get X9D+ Win64 2.8.1 release and the latest Win64 check build of this PR to crash using your procedure. Just to be sure: we are going through "MODE" (module types), aren't we?

screenshot_x9d+_23-04-10_19-46-49

**Comment by mha1** [comment]
oops, will repeat and report back

**Comment by mha1** [comment]
@elecpower Unfortunately I can confirm the crashes on 2.8.1 release, 2.9 main and this PR

**Comment by mha1** [comment]
and couldn't get Boxoer to crash too. TI tried 2.8.1 and 2.9 self build (this PR)

**Comment by elecpower** [comment]
@mha1 thank you for taking time to confirm there is a bug.
I'll test some more radio sims to see how far the issue extends.
Posting an Issue so the broader community is aware.

**Comment by elecpower** [comment]
> very strange, the X9D (no plus) doesn't crash

Tested and it crashes for me and in further testing the 128 x 64 LCD radios haven't crashed yet anyways but all 212 x 128 are crashing.

**Comment by elecpower** [comment]
Appears to be fixed by #3459. @philmoz to the rescue

**Comment by elecpower** [comment]
> Question related to thefirst bullet point: there is some code that uses define MM_RF_CUSTOM_SELECTED (multi rf protocol index 0xff). Currently multi.txt defines 93 protocols of which some are not selectable and therefore excluded in MPM's protocol scan result. 93 is the highest index (KYOSHO2) that can be selected. So currently there is no way MM_RF_CUSTOM_SELECTED (what ever protocol that is) can be selected as it has index 255. There is even code to specifically exclude MM_RF_CUSTOM_SELECTED in void MultiRfProtocols::fillBuiltinProtos():

Just had a quick check of the MPM repo and it appears to be used as an end of protocol array and unknown protocol safeguard. So for EdgeTX dropping shouldn't be a problem unless the protocol scan can return protocol 0XFF due to a bug or a MPM firmware custom self build is loaded on the module.

If it is kept I'd suggest changing the name to say MM_RF_UNKNOWN.

**Comment by elecpower** [comment]
The proposed implementation only handles known protocols. So in the situation where the pilot updates the MPM firmware to a later version than supported by ETX, how is this consistently handled by radio firmware scan, Companion and simulator especially in the area of reading, display/edit and writing model settings?

I would assume the current dummy protocols are so the unknown settings are not lost.

**Comment by elecpower** [comment]
@mha1 Hi Michael, I have no basis for deciding on RF_CUSTOM_SELECTED so leave that up to you. Agree with a block of spare protocols above the last known. As there are very few new protocols per MPM release these days then 10 spare would seem appropriate and using the default generic option also safe. Priority is not to lose/corrupt settings.

When you finish the next round of changes I will throw the kitchen sink at the code.

Keep up the excellent work.

**Comment by mha1** [comment]
Changes:
- magic sentinel number 0xfe as define
- add 10 NN protocol entries for future use of new MPM protocols not yet defined in ETX (still recommend timely ETX maintenance update if MPM releases new firmware)

@elecpower leaving RF_CUSTOM_SELECTED in doesn't do any harm apart from the next guy looking at this will scratch his head too. The well known side effects of looking at legacy code I guess. Maybe someone with recollection of the historical events that lead to the rise and fall of RF_CUSTOM_SELECTED. If better understood it should be made right or axed. For now I'll just leave it to the next generation.

I added ten future use (NN) protocols with eight possible (SUB) subtypes each and a generic setting to have all options on display and numeric general option values.

To test your use case I built fw and cpn/sim pretending to have the definitions based on an older MPM version that doesn't know KYOSHO2 (the last one of the MPM protocols that was added). Of course my 1.3.3.20 internal module knows it and also delivers it through proto scan. Selecting KYOSHO2 on the internal 1.3.3.20 MPM leads to CPN/SIM correctly displaying it as NN1/SUB1 protocol. You can see the difference in the per-defined list and proto scan if you select internal (shows KYOSHO2) and external (no KYOSHO2 but NN's) modules. The internal module doesn't deliver NN entries on proto scan which is correct. So everything seems to work as expected from my limited tests here.

Let's see what the kitchen sink can do.

PS: The best part of this exercise - not a single line of code changed. It was all done just by adding definitions in the two .h files. Basically the same exercise as adding new MPM protocols.

**Comment by mha1** [comment]
Hi Neil, Yeah, have at it, don't break it :-). I did a rebase due to two recent main merges, no functional changes.

**Comment by elecpower** [comment]
@mha1 Hi Michael, since I do not have permissions to push to your branch here is a link to mine with the extra commit https://github.com/elecpower/edgetx/tree/CpnSimMPMSingleSource

Changes:
- align field labels to radio
- change servo update frequency from combo box to spin box with edit only via up/down arrows to stop invalid values being entered via keyboard. Remove frequency list used to populate combo box as only used by Companion.
- add validation to module yaml decode checking for unsupported multi protocols and/or multi protocol sub types and if unsupported set module to OFF just to be safe.

**Comment by elecpower** [comment]
On the radio side, if the modelxx.yml file contains an unsupported/invalid multi protocol:
- the module displays a multi protocol based on uint8_t eg 277 -> Futaba.
- sub-type, the module displays the combo box but current value is [blank], if sub-types are defined for the protocol.

**Comment by elecpower** [comment]
PS I don't think I broke it

**Comment by mha1** [comment]
@elecpower: Noice, thanks Neil for your update.

> On the radio side, if the modelxx.yml file contains an unsupported/invalid multi protocol:
> 
>     * the module displays a multi protocol based on uint8_t eg 277 -> Futaba.
> 
>     * sub-type, the module displays the combo box but current value is [blank], if sub-types are defined for the protocol.

Bullet point 1: Implemented the same logic as you did for CPN on the radio side. If the RF Protocol is not in the expected range -> Module goes to OFF. Same for subtype being out of range.

Bullet point 2: I am not sure if I understood this correctly.  If yes, this is handled by bullet point 1. If not please help me to get on the same page.

**Comment by elecpower** [comment]
> Bullet point 2: I am not sure if I understood this correctly. If yes, this is handled by bullet point 1. If not please help me to get on the same page.

Handled by #1

**Comment by mha1** [comment]
@elecpower Overlooked this one, the search for 0xfe hat lots of hits. Integrated your commit. Thanks!

**Comment by elecpower** [comment]
@mha1 Hi Michael I have added a multi protocol sort to cpn as it was annoying me and with a list of ~100 protocols it is painful to find the protocol one wants. This might just do it for tweaks.
Same deal as before with extra commit on top of your last in my repo.

**Comment by mha1** [comment]
Hey Neil, the cherry on the milk shake. This is great (I am gonna miss the 1:1 index to list entry relation for testing though :-))

**Comment by elecpower** [comment]
Usability won out over testing

**Comment by mha1** [comment]
change:
- added fix for #3480: compilation problem would only show with T8 in headless config (-DRADIOMASTER_RTF_RELEASE=YES)

**Comment by mha1** [comment]
@pfeerick @elecpower Peter, Neil, are we good to go?

**Comment by elecpower** [comment]
@mha1 Michael, with reference to #2713 (and similar) the current default definition 
{MODULE_SUBTYPE_MULTI_OMP,        0, false, false,  NO_SUBTYPE,            nullptr}
IMO should be changed to 
{MODULE_SUBTYPE_MULTI_OMP,        0, false, false,  NO_SUBTYPE,            STR_MULTI_RFTUNE}
or
{MODULE_SUBTYPE_MULTI_OMP,        0, false, false,  NO_SUBTYPE,            STR_MULTI_OPTION}
so the option value is not hidden in the Companion and sim UIs when it is possible to be used.
Not ideal but we are dealing with a third party product and have to deal with its quirks.

**Comment by mha1** [comment]
Hi Neil, you are right. I finally chose STR_MULTI_RFTUNE as MPM manual refers to "Option Value:  Emulation mode and frequency fine tuning"

**Comment by mha1** [comment]
@elecpower @pfeerick good to go?

**Comment by elecpower** [comment]
Release the beast! The more testing the better.

**Comment by mha1** [comment]
@pfeerick poke. U good? This and other PR's are waiting to be merged. All reviewed and tested but in need of community testing. Let me know if I can do more of anything to get the PR's out the in the field.

**Comment by pfeerick** [comment]
LGTM... TX12-MK2 with external MPM and TX16S with internal MPM still working as expected, and Companion is a lot nicer now :)

**Comment by mha1** [comment]
Thank you!

**Pull request merged**
